src/library/scala/math/BigInt.scala

```diff
@@ -178,80 +260,173 @@ final class BigInt(val bigInteger: BigInteger)
 
   /** Compares this BigInt with the specified BigInt for equality.
    */
-  def equals (that: BigInt): Boolean = compare(that) == 0
+  def equals (that: BigInt): Boolean =
+    if (this.longEncoding)
+      that.longEncoding && (this._long == that._long)
+    else
+      !that.longEncoding && (this._bigInteger == that._bigInteger)
 
   /** Compares this BigInt with the specified BigInt
    */
-  def compare (that: BigInt): Int = this.bigInteger.compareTo(that.bigInteger)
+  def compare (that: BigInt): Int =
+    if (this.longEncoding) {
+      if (that.longEncoding) java.lang.Long.compare(this._long, that._long) else -that._bigInteger.signum()
+    } else {
+      if (that.longEncoding) _bigInteger.signum() else this._bigInteger.compareTo(that._bigInteger)
+    }
 
   /** Addition of BigInts
    */
-  def +  (that: BigInt): BigInt = new BigInt(this.bigInteger.add(that.bigInteger))
+  def +  (that: BigInt): BigInt = {
+    if (this.longEncoding && that.longEncoding) { // fast path
+      val x = this._long
+      val y = that._long
+      val z = x + y
+      if ((~(x ^ y) & (x ^ z)) >= 0L) return BigInt(z)
+    }
+    new BigInt(this.bigInteger.add(that.bigInteger))
+  }
 
   /** Subtraction of BigInts
    */
-  def -  (that: BigInt): BigInt = new BigInt(this.bigInteger.subtract(that.bigInteger))
+  def -  (that: BigInt): BigInt = {
+    if (this.longEncoding && that.longEncoding) { // fast path
+      val x = this._long
+      val y = that._long
+      val z = x - y
+      if (((x ^ y) & (x ^ z)) >= 0L) return BigInt(z)
+    }
+    new BigInt(this.bigInteger.subtract(that.bigInteger))
+  }
 
   /** Multiplication of BigInts
    */
-  def *  (that: BigInt): BigInt = new BigInt(this.bigInteger.multiply(that.bigInteger))
+  def *  (that: BigInt): BigInt = {
+    if (this.longEncoding && that.longEncoding) { // fast path
+      val x = this._long
+      val y = that._long
+      val z = x * y
+      if (x == 0 || (y == z / x && !(x == -1 && y == Long.MinValue))) return BigInt(z)
+    }
+    new BigInt(this.bigInteger.multiply(that.bigInteger))
+  }
 
   /** Division of BigInts
    */
-  def /  (that: BigInt): BigInt = new BigInt(this.bigInteger.divide(that.bigInteger))
+  def /  (that: BigInt): BigInt = {
+    if (this.longEncoding && that.longEncoding) { // fast path
+      if (this._long != Long.MinValue || that._long != -1) return BigInt(this._long / that._long)
+    }
+    new BigInt(this.bigInteger.divide(that.bigInteger))
+  }
```

Comment:
`longEncoding = _long != Long.MinValue`, so the first part of this condition is always true. so you can simplify this a bit

```suggestion
  def /  (that: BigInt): BigInt =
    if (this.longEncoding && that.longEncoding) BigInt(this._long / that._long) // fast path
    else new BigInt(this.bigInteger.divide(that.bigInteger))
```

Reply:
I don't think your suggestion is safe. If you have `Long.MinValue / -1L` you will get Long.MinValue, which is not correct:

```
scala> Long.MinValue / -1L
res1: Long = -9223372036854775808
```

The main question I have with this code, is what is the test coverage of this change. I don't think it should be merged without 100% for something so important. I'm not totally sure we have tested everything clearly. Having a scala check that BigInt is isomorphic to BigInteger for all public operations would be what I hope we can get to.

Reply:
please check the definition of `longEncoding` in the code, and see that it is precisely as I stated. thus, within the block for `if (this.longEncoding && ...)`, it must be the case that `_long != Long.MinValue`. in which case checking that same condition again is redundant

Reply:
we could certainly add a comment noting that `this._long != Long.MinValue` and therefore no other checks are needed, for clarity

Reply:
Indeed, you are correct, I was mistaken. I agree the comment would help future readers not make my same mistake.

Reply:
I've simplified the code in the direction suggested, and put the original code as a comment.